File: Data Structures & Algorithms/combination-target-sum/submission-0.py
class Solution:
    def combinationSum(self, nums: List[int], target: int) -> List[List[int]]:
        res = []
        def backtrack(start, path, total):
            # base case
            if total == target:
                res.append(path.copy()) # snap shot of the current path
                return

            # path cannot work -> stop exploring
            if total > target:
                return
            
            # keep adding new numbers to combination
            for i in range(start, len(nums)):
                # add the current number to the path
                path.append(nums[i])
                # go deep 
                backtrack(i, path, total + nums[i])
                # pop the path to backtrack
                path.pop()
        backtrack(0, [], 0)
        return res

File: Data Structures & Algorithms/remove-duplicates-from-sorted-array-ii/submission-0.ts
class Solution {
    /**
     * @param {number[]} nums
     * @return {number}
     */
    removeDuplicates(nums: number[]): number {
        let s =0;
        let count = 1;
        for( let f = 1; f<nums.length; f++){
            if(nums[f] === nums[f-1])count++; 
            if ( nums[f] === nums[f-1] && nums[f] === nums[f-2] && count > 2) continue;
            if(nums[f] !== nums[f-1]) count = 1;
            s++;
            nums[s] = nums[f];
        }
        return s+1;
    }
}